conmon-rs/server/src/child_reaper.rs

```diff
@@ -0,0 +1,124 @@
+//! Child process reaping and management.
+
+use anyhow::{bail, Result};
+use getset::Getters;
+use log::{debug, error};
+use nix::sys::{
+    signal::{kill, Signal, SIGCHLD},
+    signalfd::{siginfo, SfdFlags, SigSet, SignalFd},
+};
+use nix::unistd::Pid;
+use std::path::PathBuf;
+use std::thread;
+use std::{collections::HashMap, fs::File, io::Write, sync::Arc, sync::Mutex};
+
+impl ChildReaper {
+    pub fn init_reaper(&self) -> Result<()> {
+        let mut mask = SigSet::empty();
+        mask.add(SIGCHLD);
+        mask.thread_block()?;
+        // Signals are queued up on the file descriptor
+        let mut sfd = SignalFd::with_flags(&mask, SfdFlags::empty())?;
+
+        let children = self.children.clone();
+        thread::spawn(move || {
+            let children = children.clone();
+            loop {
+                let children = children.clone();
+                match sfd.read_signal() {
+                    // we caught a signal
+                    Ok(Some(sig)) => {
+                        // Immediately spawn a thread to reduce risk of dropping
+                        // a SIGCHLD.
```

Comment:
Ah...wow yeah, I think using pidfd would be cleaner.

Reply:
one day..

Reply:
https://github.com/tokio-rs/tokio/issues/4016